### ISSUE TITLE
daemon becomes unresponsive after sustained load

### ISSUE THREAD

**Issue by amidvidy** [moby/moby#33447]
**Description**

We have a cluster of Centos7 machines that runs short batch jobs via `docker run` invocations. We are seeing the docker daemon 'hang' after about 1-2 weeks of sustained job load. 

When the daemon is in a 'hung' state, for any containers that were previously running, the parent `docker run` invocation will not exit, even if the process started  in the `ENTRYPOINT` completes. Attempting to run a new container will hang as well. The only solution is to restart the daemon.

A typical healthcheck would be to run:

`timeout 3 sudo docker run <base linux image> /bin/true`

**Steps to reproduce the issue:**

It's difficult to reproduce. Here is a vague way it potentially could be reproduced.

1) Create a docker image that runs python, downloads some data, processes it, and saves it somewhere.

2) Run # CPU cores of this image concurrently in a loop using `docker run` until the daemon becomes unresponsive.
 
**Describe the results you received:**

Containers failing to start & exit due to unresponsive daemon.

**Describe the results you expected:**

Containers starting & exiting as expected.

**Additional information you deem important (e.g. issue happens only occasionally):**

Issue seems to manifest at least once a day on a cluster of 15 machines under moderate to heavy load. At the time the daemon goes unresponsive, it will typically have plenty of memory available and not be maxing out IOPS or anything else that suggests resource exhaustion.

**Output of `docker version`:**

```
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:38:28 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:38:28 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

**Output of `docker info`:**

```
Containers: 11
 Running: 1
 Paused: 0
 Stopped: 10
Images: 329
Server Version: 1.13.1
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.6.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.01 GiB
Name: core-prod7
ID: QMJD:HSXK:OIDI:D373:SP32:RPKG:CGUB:5HXR:PNL7:5445:CZUW:Y3JO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
amidvidy@core-prod7:~$ 

```
Here are some suspicious log messages from the daemon log. At the time the daemon was in an unresponsive state and I had another shell running `sudo docker run <base linux image> /bin/true`. 

```
May 31 07:10:08 core-prod7 dockerd[20200]: time="2017-05-31T07:10:08.052904543Z" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container d3890c56e50324155266d23300355a479aa6cb2bd3869d57217d5008b21a7d5f: rpc error: code = 2 desc = containerd: container not found"
May 31 07:10:10 core-prod7 dockerd[20200]: time="2017-05-31T07:10:10.131000244Z" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container d3890c56e50324155266d23300355a479aa6cb2bd3869d57217d5008b21a7d5f: rpc error: code = 2 desc = containerd: container not found"
May 31 07:10:12 core-prod7 dockerd[20200]: time="2017-05-31T07:10:12.611300498Z" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container d3890c56e50324155266d23300355a479aa6cb2bd3869d57217d5008b21a7d5f: rpc error: code = 2 desc = containerd: container not found"
May 31 07:10:12 core-prod7 dockerd[20200]: time="2017-05-31T07:10:12.785139546Z" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container d3890c56e50324155266d23300355a479aa6cb2bd3869d57217d5008b21a7d5f: rpc error: code = 2 desc = containerd: container not found"
May 31 07:10:12 core-prod7 dockerd[20200]: time="2017-05-31T07:10:12.956441841Z" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container d3890c56e50324155266d23300355a479aa6cb2bd3869d57217d5008b21a7d5f: rpc error: code = 2 desc = containerd: container not found"
May 31 07:10:15 core-prod7 dockerd[20200]: time="2017-05-31T07:10:15.591229486Z" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container d3890c56e50324155266d23300355a479aa6cb2bd3869d57217d5008b21a7d5f: rpc error: code = 2 desc = containerd: container not found"
May 31 07:10:38 core-prod7 dockerd[20200]: time="2017-05-31T07:10:38.796109889Z" level=info msg="goroutine stacks written to /var/run/docker/goroutine-stacks-2017-05-31T071038Z.log"
```

Here is a SIGUSR1 stack dump taken when the daemon was in an unresponsive state:

https://gist.github.com/amidvidy/3b284bcd65c54d55465b274c8231a68c

Please let me know if there is any other information I can provide. The issue happens frequently enough that I can run diagnostics on a box when it happens.

**Comment by amidvidy** [comment]
Let me clarify...

`docker ps` works (will verify the next time it happens). However any invocations of `docker run` will block indefinitely, despite the machine having plenty of memory/cpu available. Additionally any `docker run` invocations that were started before the daemon became 'hung' will not actually complete, even if the process within the container finishes.

**Comment by amidvidy** [comment]
potentially related to #24985

**Comment by amidvidy** [comment]
also potentially related to #25221

**Comment by cpuguy83** [comment]
Unfortunately I'm just not seeing anything in the stack dump that would indicate a problem.
Do you have any logs from the daemon? Kernel messages while the daemon is unresponsive?

**Comment by amidvidy** [comment]
Here are the daemon logs from a few hours before the event to the point at which I restarted the daemon via systemctl.


[docker-hang-log.txt](https://github.com/moby/moby/files/1042840/docker-hang-log.txt)

**Comment by thaJeztah** [comment]
@amidvidy the actual dumps are written to separate files; can you attach those? See the output of your logs;

```
goroutine stacks written to /var/run/docker/goroutine-stacks-2017-05-31T071038Z.log
daemon datastructure dump written to /var/run/docker/daemon-data-2017-05-31T071038Z.log
```

**Comment by thaJeztah** [comment]
Part of the problem  may be related to;

```
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: false
```

Running overlay (or overlay2) on XFS without `d_type` is known to cause issues on CentOS and RHEL. Docker currently "allows" it for backward-compatibility, but will block it in future, you'll see a warning being logged about this as well when starting the daemon;

```
overlay: the backing xfs filesystem is formatted without d_type support,
which leads to incorrect behavior. Reformat the filesystem with ftype=1
to enable d_type support.
Running without d_type support will no longer be supported in Docker 17.12.
```

Also see these issues, which contain some information https://github.com/moby/moby/issues/10294, https://github.com/moby/moby/issues/29873.

The RHEL release notes mentions this; https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/7.2_Release_Notes/technology-preview-file_systems.html

> Note that XFS file systems must be created with the `-n ftype=1` option enabled for use as an overlay. With the rootfs and any file systems created during system installation, set the `--mkfsoptions=-n ftype=1` parameters in the Anaconda kickstart. When creating a new file system after the installation, run the `# mkfs -t xfs -n ftype=1 /PATH/TO/DEVICE` command. To determine whether an existing file system is eligible for use as an overlay, run the `# xfs_info /PATH/TO/DEVICE | grep ftype` command to see if the `ftype=1` option is enabled.
> Note that OverlayFS provides a restricted set of the POSIX standards. Test your application thoroughly before deploying it with OverlayFS.

**Comment by amidvidy** [comment]
Thanks @thaJeztah. See attached zip file containing the 2 files. I have lightly redacted the daemon dump to not include the names of our internal images.

[docker-diagnostics.zip](https://github.com/moby/moby/files/1045438/docker-diagnostics.zip)

**Comment by cpuguy83** [comment]
@amidvidy Do you have the output of `dmesg` while this is happening?
The stack dump is not really revealing anything wrong from within Docker.
Everything appears to be running fine within the daemon itself.

The actual daemon logs from this time period, especially with debug enabled, would be helpful as well.

Thanks! Sorry about the back and forth for more info.

**Comment by amidvidy** [comment]
Yes, the issue reoccured this morning. I didn't have debug enabled on the daemon (will do so now for this host). Here is the dmesg output.

1) In one terminal pane I run this command:

```
centos@core-prod8:~$ sudo docker run docker.eglp.com/eglp/centos7 /bin/bash -c  'echo hello'                                                                                                                    
hello
```
Docker run invocation does not exit despite the bash process in the container completing.

2) In another terminal pane I ran `dmesg --follow`. Here are the new messages that appeared after I ran the command:

```
[8363391.914390] overlayfs: upper fs needs to support d_type. This is an invalid configuration.
[8363391.927090] overlayfs: upper fs needs to support d_type. This is an invalid configuration.
[8363391.938593] overlayfs: upper fs needs to support d_type. This is an invalid configuration.
[8363391.943860] device vethff25580 entered promiscuous mode
[8363391.947224] IPv6: ADDRCONF(NETDEV_UP): vethff25580: link is not ready
[8363391.951153] docker0: port 7(vethff25580) entered forwarding state
[8363391.954673] docker0: port 7(vethff25580) entered forwarding state
[8363391.958537] docker0: port 7(vethff25580) entered disabled state
[8363392.038453] IPv6: ADDRCONF(NETDEV_CHANGE): vethff25580: link becomes ready
[8363392.042465] docker0: port 7(vethff25580) entered forwarding state
[8363392.046194] docker0: port 7(vethff25580) entered forwarding state
[8363392.116768] docker0: port 7(vethff25580) entered disabled state
[8363392.122780] docker0: port 7(vethff25580) entered disabled state
[8363392.129124] device vethff25580 left promiscuous mode
[8363392.132228] docker0: port 7(vethff25580) entered disabled state
[8363404.287126] overlayfs: upper fs needs to support d_type. This is an invalid configuration.
[8363404.305220] overlayfs: upper fs needs to support d_type. This is an invalid configuration.
[8363404.316986] overlayfs: upper fs needs to support d_type. This is an invalid configuration.
[8363404.322392] device veth368d5fb entered promiscuous mode
[8363404.325782] IPv6: ADDRCONF(NETDEV_UP): veth368d5fb: link is not ready
[8363404.329406] docker0: port 7(veth368d5fb) entered forwarding state
[8363404.333040] docker0: port 7(veth368d5fb) entered forwarding state
[8363404.336688] docker0: port 7(veth368d5fb) entered disabled state
[8363404.411697] IPv6: ADDRCONF(NETDEV_CHANGE): veth368d5fb: link becomes ready
[8363404.415728] docker0: port 7(veth368d5fb) entered forwarding state
[8363404.419536] docker0: port 7(veth368d5fb) entered forwarding state
[8363404.500365] docker0: port 7(veth368d5fb) entered disabled state
[8363404.507448] docker0: port 7(veth368d5fb) entered disabled state
[8363404.513851] device veth368d5fb left promiscuous mode
[8363404.516960] docker0: port 7(veth368d5fb) entered disabled state
[8363446.087915] overlayfs: upper fs needs to support d_type. This is an invalid configuration.
[8363446.099008] overlayfs: upper fs needs to support d_type. This is an invalid configuration.
[8363446.113216] overlayfs: upper fs needs to support d_type. This is an invalid configuration.
[8363446.118333] device veth359700e entered promiscuous mode
[8363446.121596] IPv6: ADDRCONF(NETDEV_UP): veth359700e: link is not ready
[8363446.125350] docker0: port 7(veth359700e) entered forwarding state
[8363446.128902] docker0: port 7(veth359700e) entered forwarding state
[8363446.132752] docker0: port 7(veth359700e) entered disabled state
[8363446.201654] IPv6: ADDRCONF(NETDEV_CHANGE): veth359700e: link becomes ready
[8363446.205605] docker0: port 7(veth359700e) entered forwarding state
[8363446.209373] docker0: port 7(veth359700e) entered forwarding state
```

**Comment by cpuguy83** [comment]
> overlayfs: upper fs needs to support d_type. This is an invalid configuration.

I suspect that needs to be resolved before we can go much further as it is likely the culprit.

**Comment by amidvidy** [comment]
thanks - we're actually doing some testing on both `overlay` on a newer xfs that supports `dtype=true` as well as `devicemapper` on `direct-lvm` in preparation for a migration. Feel free to close out, we will open a new ticket if the problem persists after upgrading.

**Comment by thaJeztah** [comment]
@amidvidy note that the newer CentOS7 kernels also support the features to use the `overlay2` driver. Docker is unable to detect it because the feature was backported (and there's no way to feature-detect them); You can skip the kernel-version detection (which allows `overlay2` to be used), by setting the [`overlay2.override_kernel_check=1` storage driver option](https://github.com/moby/moby/blob/67fdf574d5acd6ddccb6ece0ffe0ace1c1608712/docs/reference/commandline/dockerd.md#overlay2-options)

**Comment by thaJeztah** [comment]
Let me close this issue for now, following the discussion above 👍
